x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/categorization/TreeNode.java

```diff
@@ -0,0 +1,404 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.aggs.categorization;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.util.Accountable;
+import org.elasticsearch.core.Tuple;
+import org.elasticsearch.search.aggregations.AggregationExecutionException;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.stream.Collectors;
+
+import static org.apache.lucene.util.RamUsageEstimator.NUM_BYTES_OBJECT_REF;
+import static org.apache.lucene.util.RamUsageEstimator.sizeOfCollection;
+import static org.apache.lucene.util.RamUsageEstimator.sizeOfMap;
+import static org.elasticsearch.xpack.ml.aggs.categorization.CategorizationBytesRefHash.WILD_CARD_ID;
+
+/**
+ * Tree node classes for the categorization token tree.
+ *
+ * Two major node types exist:
+ *  - Inner: which are nodes that have children token nodes
+ *  - Leaf: Which collection multiple {@link TextCategorization} based on similarity restrictions
+ */
+abstract class TreeNode implements Accountable {
+
+    private static final Logger LOGGER = LogManager.getLogger(TreeNode.class);
+
+    private long count;
+
+    TreeNode(long count) {
+        this.count = count;
+    }
+
+    abstract void mergeWith(TreeNode otherNode);
+
+    abstract boolean isLeaf();
+
+    final void incCount(long count) {
+        this.count += count;
+    }
+
+    final long getCount() {
+        return count;
+    }
+
+    // TODO add option for calculating the cost of adding the new group
+    abstract TextCategorization addLog(int[] logTokenIds, long docCount, TreeNodeFactory treeNodeFactory);
+
+    abstract TextCategorization getLogGroup(int[] logTokens);
+
+    abstract List<TextCategorization> getAllChildrenLogGroups();
+
+    abstract void collapseTinyChildren();
+
+    static class LeafTreeNode extends TreeNode {
+        private final List<TextCategorization> textCategorizations;
+        private final int similarityThreshold;
+
+        LeafTreeNode(long count, int similarityThreshold) {
+            super(count);
+            this.textCategorizations = new ArrayList<>();
+            this.similarityThreshold = similarityThreshold;
+            if (similarityThreshold < 1 || similarityThreshold > 100) {
+                throw new IllegalArgumentException("similarityThreshold must be between 1 and 100");
+            }
+        }
+
+        public boolean isLeaf() {
```

Comment:
Could you add `@Override` here? It just helps me speed read this stuff when it's there.

Reply:
yeah, thats a mistake on my part, it should be an `@Override`